modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -1206,28 +1208,87 @@ public int partitionId(BinaryRowEx row) {
     }
 
     /**
-     * TODO asch keep the same order as for keys Collects multirow responses from multiple futures into a single collection IGNITE-16004.
+     * Gathers the result of batch processing into a single resulting collection of rows.
      *
-     * @param futs Futures.
-     * @return Row collection.
+     * @param rowBatchByPartitionId Row batches by partition ID.
+     * @return Future of collecting results.
      */
-    private CompletableFuture<Collection<BinaryRow>> collectMultiRowsResponses(CompletableFuture<Object>[] futs) {
-        return CompletableFuture.allOf(futs)
+    static CompletableFuture<Collection<BinaryRow>> collectMultiRowsResponsesWithoutRestoreOrder(
+            Int2ObjectMap<RowBatch> rowBatchByPartitionId
+    ) {
+        return allOfResultFuture(rowBatchByPartitionId)
                 .thenApply(response -> {
-                    Collection<BinaryRow> list = new ArrayList<>(futs.length);
+                    var result = new ArrayList<BinaryRow>(rowBatchByPartitionId.size());
 
-                    for (CompletableFuture<Object> future : futs) {
-                        Collection<BinaryRow> values = (Collection<BinaryRow>) future.join();
+                    for (Int2ObjectMap.Entry<RowBatch> partitionRowBatch : rowBatchByPartitionId.int2ObjectEntrySet()) {
+                        Collection<BinaryRow> batchResult = getBatchProcessingResult(partitionRowBatch);
```

Comment:
What's the reason of passing entry into this method, instead of `RowBatch`? Does this method require partition ID?

Reply:
fix it